README.md
# Jfrog::Saas::Log::Collector

JFrog Saas Log Collector gem is intended for downloading and extracting of log files generated in Artifactory or Xray on the Jfrog Cloud.
The Log Collection feature on the cloud instance has to be [enabled](https://www.jfrog.com/confluence/display/JFROG/Artifactory+REST+API#ArtifactoryRESTAPI-EnableLogCollection) for this gem to perform the download and extract of the logs.

## Installation

Prerequisite: 

For Gem based install, Ruby Interpreter has to be setup first, following is the recommended process to install Ruby

```text

1. Install Ruby Version Manager (RVM) as described in https://rvm.io/rvm/install#installation-explained, ensure to follow all the onscreen instructions provided to complete the rvm installation
	* For installation across users a SUDO based install is recommended, the installation is as described in https://rvm.io/support/troubleshooting#sudo

2. Once rvm installation is complete, verify the RVM installation executing the command 'rvm -v'

3. Now install ruby v2.7.0 or above executing the command 'rvm install <ver_num>', ex: 'rvm install 2.7.5'

4. Verify the ruby installation, execute 'ruby -v', gem installation 'gem -v' and 'bundler -v' to ensure all the components are intact

5. Post completion of Ruby, Gems installation, the environment is ready to further install new gems, execute the following gem install commands one after other to setup the needed ecosystem

```
Install it:

    $ gem install jfrog-saas-log-collector

Or add this line to your application's Gemfile:

```ruby
gem 'jfrog-saas-log-collector'
```

And then execute:

    $ bundle install

## Usage

Once the gem is successfully installed, use the gem to generate the sample config file,

    $ jfrog-saas-log-collector -g <full_path_of_the_config_file>

example

    $  jfrog-saas-log-collector -g /var/opt/jfrog/saas/sampleconfig.yaml

The config file sample would look like this, 

```yaml
connection:
  jpd_url: "<saas_jpd_url>"
  end_point_base: "artifactory"
  username: "<admin_user>"
  access_token: "<admin_access_token>"
  open_timeout_in_secs: 20
  read_timeout_in_secs: 60
log:
  log_ship_config: "access/api/v1/logshipping/config"
  solutions_enabled: ["artifactory", "xray"]
  log_types_enabled: ["access-request", "router-request"]
  uri_date_pattern: "%Y-%m-%d"
  audit_repo: "artifactory/jfrog-logs-audit"
  log_repo: "artifactory/jfrog-logs"
  debug_mode: false
  target_log_path: "<path_to_extract_logs_for_artifactory>"
  print_with_utc: false
  log_file_retention_days: 5
process:
  parallel_process: 2
  parallel_downloads: 5
  historical_log_days: 2
  write_logs_by_type: false
  minutes_between_runs: 180
```
Provide all the necessary values to the tags which have angular braces like "<saas_jpd_url>" to "https://example.jfrog.io", fill in for all other segments. Do not change any other values unless the operation associated with the other tag is understood well.
Once done, the jfrog-saas-log collector execution can be started by executing the command

    $ jfrog-saas-log-collector -c <full_path_of_the_config_file>

example

    $  jfrog-saas-log-collector -c /var/opt/jfrog/saas/sampleconfig.yaml

The gem records the progress or errors on the STDOUT / console and in a logfile (if the config provided is valid) which can be located at  <path_to_extract_logs_for_artifactory>/jfrog-saas-collector.log

For all the options supported use

    $ jfrog-saas-log-collector -h      OR    $ jfrog-saas-log-collector --help

```shell
Usage: jfrog-saas-log-collector [options]
    -h, --help                       Prints this help
    -g, --generate=CONFIG            Generates sample config file from template to target file provided
    -c, --config=CONFIG              Executes the jfrog-saas-log-collector with the config file provided
```

## Contributing

Bug reports are welcome on GitHub at https://github.com/jfrog/jfrog-saas-log-collector/issues.



CHANGELOG.md
## [ jfrog-saas-log-collector change log ]

## [0.1.1] - 2022-04-04

- Initial release

## [0.1.1 - 0.1.4] - 2022-04-05 - Yanked

- Added Scheduler and Job concurrency handling if the gem is run in multiple terminals for the same config

## [0.1.5] - 2022-04-06

- Fixes to config file validation due to missing encoding


lib/jfrog/saas/log/confighandler.rb
# frozen_string_literal: true

require 'optparse'
require 'yaml'
require 'singleton'
require 'fileutils'

require_relative 'commonutils'
require_relative 'constants'

module Jfrog
  module Saas
    module Log
      class LogConfig
        include Singleton

        def self.logger
          logger
        end

        def self.console_logger
          console_logger
        end

        def self.log_ship_config
          log_ship_config
        end

        def self.solutions_enabled
          solutions_enabled
        end

        def self.log_types_enabled
          log_types_enabled
        end

        def self.uri_date_pattern
          uri_date_pattern
        end

        def self.audit_repo_url
          audit_repo_url
        end

        def self.log_repo_url
          log_repo_url
        end

        def self.debug_mode
          debug_mode
        end

        def self.target_log_path
          target_log_path
        end

        def self.print_with_utc
          print_with_utc
        end

        def self.log_file_retention_days
          log_file_retention_days
        end

        attr_accessor :solutions_enabled, :log_types_enabled, :uri_date_pattern, :audit_repo_url, :log_repo_url, :debug_mode, :target_log_path, :log_ship_config, :logger, :console_logger, :print_with_utc, :log_file_retention_days

        def initialize; end

        def configure(config_file)
          config = YAML.load_file(config_file)
          self.target_log_path = (config['log']['target_log_path']).to_s.strip
          log_file = "#{target_log_path}/jfrog-saas-collector.log"
          FileUtils.mkdir_p(target_log_path) unless File.directory? target_log_path
          FileUtils.touch(log_file) unless File.exist? log_file
          self.logger = Logger.new(log_file, 'weekly')
          self.console_logger = Logger.new($stdout)

          console_logger.formatter = logger.formatter = proc do |severity, datetime, progname, msg|
            formatted_date = datetime.strftime('%Y-%m-%d %H:%M:%S')
            "[ #{formatted_date}, p_id=##{Process.pid}, t_id=##{Thread.current.object_id}, #{severity.ljust(5)}] -- #{msg} \n"
          end

          self.log_ship_config = (config['log']['log_ship_config'])
          self.solutions_enabled = config['log']['solutions_enabled'].map(&:strip)
          self.log_types_enabled = config['log']['log_types_enabled'].map(&:strip)
          self.uri_date_pattern = (config['log']['uri_date_pattern']).to_s
          self.audit_repo_url = (config['log']['audit_repo']).to_s.strip
          self.log_repo_url = (config['log']['log_repo']).to_s.strip
          self.debug_mode = (config['log']['debug_mode'])
          self.print_with_utc = (config['log']['print_with_utc'])
          self.log_file_retention_days = if (config['log']['log_file_retention_days']).to_i.positive?
                                           (config['log']['log_file_retention_days']).to_i
                                         else
                                           7
                                         end
        end

        def to_s
          "Object_id :#{object_id}, solutions_enabled :#{solutions_enabled}, log_types_enabled:#{log_types_enabled}, uri_date_pattern: #{uri_date_pattern}"
        end
      end

      class ConnectionConfig
        include Singleton

        def self.jpd_url
          jpd_url
        end

        def self.end_point_base
          end_point_base
        end

        def self.username
          username
        end

        def self.access_token
          access_token
        end

        def self.api_key
          api_key
        end

        def self.ignore_errors_in_response
          ignore_errors_in_response
        end

        def self.open_timeout_in_secs
          open_timeout_in_secs
        end

        def self.read_timeout_in_secs
          read_timeout_in_secs
        end

        attr_accessor :jpd_url, :end_point_base, :username, :access_token, :api_key, :ignore_errors_in_response, :open_timeout_in_secs, :read_timeout_in_secs

        def initialize; end

        def configure(config_file)
          config = YAML.load_file(config_file)
          self.jpd_url = (config['connection']['jpd_url']).to_s.strip
          self.username = (config['connection']['username']).to_s.strip
          self.access_token = (config['connection']['access_token']).to_s.strip
          self.end_point_base = (config['connection']['end_point_base']).to_s.strip
          self.api_key = (config['connection']['api_key']).to_s.strip
          self.ignore_errors_in_response = (config['connection']['ignore_errors_in_response'])

          self.open_timeout_in_secs = if config['connection']['open_timeout_in_secs'].to_i.positive?
                                        config['connection']['open_timeout_in_secs'].to_i
                                      else
                                        5
                                      end

          self.read_timeout_in_secs = if config['connection']['read_timeout_in_secs'].to_i.positive?
                                        config['connection']['read_timeout_in_secs'].to_i
                                      else
                                        60
                                      end
        end

        def to_s
          "Object_id :#{object_id}, jpd_url :#{jpd_url}, username:#{username}, access_token: #{access_token}, end_point_base:#{end_point_base}, api_key:#{api_key}"
        end
      end

      class ProcessConfig
        include Singleton

        def self.parallel_process
          parallel_process
        end

        def self.parallel_downloads
          parallel_downloads
        end

        def self.historical_log_days
          historical_log_days
        end

        def self.write_logs_by_type
          write_logs_by_type
        end

        def self.minutes_between_runs
          minutes_between_runs
        end

        attr_accessor :parallel_process, :parallel_downloads, :historical_log_days, :write_logs_by_type, :minutes_between_runs

        def initialize; end

        def configure(config_file)
          config = YAML.load_file(config_file)
          self.parallel_downloads = if config['process']['parallel_process'].to_i.positive?
                                      config['process']['parallel_process'].to_i
                                    else
                                      1
                                    end

          self.parallel_downloads = if config['process']['parallel_downloads'].to_i.positive?
                                      config['process']['parallel_downloads'].to_i
                                    else
                                      1
                                    end

          self.historical_log_days = if config['process']['historical_log_days'].to_i.positive?
                                       config['process']['historical_log_days'].to_i
                                     else
                                       1
                                     end

          self.write_logs_by_type = config['process']['write_logs_by_type']

          self.minutes_between_runs = if config['process']['minutes_between_runs'].to_i.positive?
                                        config['process']['minutes_between_runs'].to_i
                                      else
                                        180
                                      end
        end

        def to_s
          "Object_id :#{object_id}, parallel_downloads: #{parallel_downloads}, historical_log_days: #{historical_log_days}, write_logs_by_type: #{write_logs_by_type}"
        end
      end

      class ConfigHandler
        include Singleton
        @conn_config = nil
        @log_config = nil
        @proc_config = nil
        @config_path = nil

        @@file_name = ''

        def self.file_name(file_name)
          @@file_name = file_name
        end

        class << self
          attr_reader :conn_config
        end

        class << self
          attr_reader :log_config
        end

        class << self
          attr_reader :proc_config
        end

        attr_accessor :conn_config, :log_config, :proc_config, :file_name

        def initialize
          @mutex = Mutex.new
          # If not found in options, check for the environment variable
          @config_path = if !@@file_name.nil?
                           @@file_name
                         else
                           ENV['LOG_COLLECTOR_CONFIG']
                         end
          load_all_config(@config_path, 'initialize')
        end

        def load_all_config(config_file, thread_name)
          @mutex.synchronize do
            if !config_file.nil?
              @log_config = LogConfig.instance
              @log_config.configure(config_file)
              @conn_config = ConnectionConfig.instance
              @conn_config.configure(config_file)
              @proc_config = ProcessConfig.instance
              @proc_config.configure(config_file)

              MessageUtils.instance.log_message(MessageUtils::CONFIG_LOAD_BEGIN, { "param1": thread_name, "param2": config_file, "#{MessageUtils::LOG_LEVEL}": CommonUtils::LOG_INFO })
              if LogConfig.instance.debug_mode == true
                MessageUtils.instance.log_message(MessageUtils::CONFIG_LOAD_DETAIL, { "param1": thread_name, "param2": @conn_config.to_s, "#{MessageUtils::LOG_LEVEL}": CommonUtils::LOG_DEBUG })
              end
              if LogConfig.instance.debug_mode == true
                MessageUtils.instance.log_message(MessageUtils::CONFIG_LOAD_DETAIL, { "param1": thread_name, "param2": @log_config.to_s, "#{MessageUtils::LOG_LEVEL}": CommonUtils::LOG_DEBUG })
              end
              if LogConfig.instance.debug_mode == true
                MessageUtils.instance.log_message(MessageUtils::CONFIG_LOAD_DETAIL, { "param1": thread_name, "param2": @proc_config.to_s, "#{MessageUtils::LOG_LEVEL}": CommonUtils::LOG_DEBUG })
              end
              MessageUtils.instance.log_message(MessageUtils::CONFIG_LOAD_END, { "param1": thread_name, "param2": config_file, "#{MessageUtils::LOG_LEVEL}": CommonUtils::LOG_INFO })
            else
              MessageUtils.instance.log_message(MessageUtils::CONFIG_ERROR_NO_FILE, { "#{MessageUtils::LOG_LEVEL}": CommonUtils::LOG_ERROR })
            end
          end
        end
      end
    end
  end
end
